timestamp/src/token2/services.rs

```diff
@@ -0,0 +1,108 @@
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use super::state_machine::*;
+use super::types::*;
+use super::ServiceHandler;
+pub use ckey::{Ed25519Private as Private, Ed25519Public as Public};
+use coordinator::module2::*;
+use coordinator::types::*;
+use coordinator::{Header, Transaction};
+use primitives::H256;
+use remote_trait_object::{service, Service};
+use std::collections::BTreeSet;
+
+#[service]
+pub trait TokenManager: Service {
+    // Immutable accesses
+    fn get_account(&self, session: SessionKey, public: &Public, default: bool) -> Result<Account, Error>;
+    fn get_owning_accounts_with_issuer(&self, session: SessionKey, issuer: &H256) -> Result<BTreeSet<Public>, Error>;
+
+    // Mutable accesses
+    fn issue_token(&self, session: SessionKey, issuer: &H256, receiver: &Public) -> Result<(), Error>;
+}
+
+impl TokenManager for ServiceHandler {
+    fn get_account(&self, session: SessionKey, public: &Public, default: bool) -> Result<Account, Error> {
+        let state_machine = self.create_state_machine(session);
+        state_machine.execute_access(GetAccount {
+            public,
+            default,
+        })
+    }
+
+    fn get_owning_accounts_with_issuer(&self, session: SessionKey, issuer: &H256) -> Result<BTreeSet<Public>, Error> {
+        let state_machine = self.create_state_machine(session);
+        state_machine.execute_access(GetOwningAccountsWithIssuer {
+            issuer,
+        })
+    }
+
+    fn issue_token(&self, session: SessionKey, issuer: &H256, receiver: &Public) -> Result<(), Error> {
+        let state_machine = self.create_state_machine(session);
+        state_machine.execute_transition(IssueToken {
+            issuer,
+            receiver,
+        })
+    }
+}
+
+impl TxOwner for ServiceHandler {
+    fn block_opened(&self, _session: SessionKey, _header: &Header) -> Result<(), HeaderError> {
+        Ok(())
+    }
+
+    fn execute_transaction(&self, session: SessionKey, transaction: &Transaction) -> Result<TransactionOutcome, ()> {
+        let state_machine = self.create_state_machine(session);
+
+        let get_sequence =
+            |public: &Public| self.account_manager.read().get_account(session, public, true).map(|x| x.seq);
+        let increase_sequence = move |public: &Public| {
+            self.account_manager.read().increase_sequence(session, public, true).unwrap();
```

Comment:
What is the benefit of creating these functions instead of passing account_manager?
If we keep this way, we need to write every function of other modules that is used by the transaction. It is not a scalable way and needs extra work.

Reply:
1. The state machine doesn't have to know the concept of session.
2. The state machine doesn't have to know the presence of services.
3. The state machine will be provided only required methods.

Reply:
I agree with number 1. Managing sessions is not the role of the state machine.
I'm not sure about the number 3.

I don't agree with the number 2.
State machine needs to know the service in two aspects. The first one is about locking. Calling service is a blocking call. If state machines is locking some resource, it is better to avoid the blocking call to reduce the lock region. The second one is about data consistency. If state machine read data from the state before calling service, it can be changed after the service call. The state machine carefully order calls of service methods.

Reply:
We conclude to seek better way later.